### DISCUSSION TITLE
Tutorial on how to create custom components

### DISCUSSION THREAD

**Discussion by bph** [WordPress/developer-blog-content#252] [Topic Ideas]
In the outreach channel, there was a [discussion around this topic](https://wordpress.slack.com/archives/C015GUFFC00/p1713864175741239): 

Jessica Lyschik  elaborated: 
> When creating a custom block, I have some settings defined inside InspectorControls with like toggles, selects, inputs, etc. Now I have to implement a setting that I can't get done with the provided Components, so I'd like to know how to create my own component, preferably in JSX.

Fabian Kägy had some tips: 

> I tend to start by wrapping my control in the [BaseControl](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/base-control)
>
>It already gives you all the styles for labels, help text etc. And from there it really is just plain react, and you can do whatever you want.
>
>I’m building components like that here in this package for example: https://github.com/10up/block-components/
>
>This here is a simple one that just exports a light wrapper around another control: https://github.com/10up/block-components/blob/develop/components/color-settings/index.js
>
>But it can also be much more complex like in this example here for an icon picker: https://github.com/10up/block-components/blob/88f68d8df1e3912abf4c976b7c8b8fa72b964a3d/components/icon-picker/icon-picker.js#L68-L91

**Comment by fabiankaegy** [comment]
Love this idea and would be happy to create a writeup for how to create these custom components 👍

**Reply by luminuu** [reply]
Ping me if you need someone to proof read.

**Comment by bph** [comment]
This topic was approved in today's meeting. The conversation continues on the issue 
- #257

**Comment by juanmaguitar** [comment]
A nice addition to this post could be how to enhance core components with custom hooks and higher order components ([here's](https://github.com/WordPress/block-development-examples/tree/trunk/plugins/editable-block-1b8c51) an example - see [`useCharLimit`](https://github.com/WordPress/block-development-examples/blob/trunk/plugins/editable-block-1b8c51/src/blocks/enhanced-editable-block-1b8c51/js/useCharLimit.js) and [`withCharLimit`](https://github.com/WordPress/block-development-examples/blob/trunk/plugins/editable-block-1b8c51/src/blocks/enhanced-editable-block-1b8c51/js/withCharLimit.js))
Or maybe this is material for a complementary post (that I'd be happy to write about).

**Reply by bph** [reply]
The discussion should move to the issue @juanmaguitar as this discussion is locked for non-maintainers.

**Reply by juanmaguitar** [reply]
Ups! I didn't realize that. 
Thanks for the heads up! 🙏
